tests/Unit/Elliptic/Systems/Poisson/Actions/Test_Observe.cpp

```diff
@@ -0,0 +1,302 @@
+      typename MockRuntimeSystem::template MockDistributedObjectsTag<
+          element_comp>;
+  TupleOfMockDistributedObjects dist_objects{};
+  tuples::get<ObserverMockDistributedObjectsTag>(dist_objects)
+      .emplace(0, ActionTesting::MockDistributedObject<obs_component>{});
+  tuples::get<WriterMockDistributedObjectsTag>(dist_objects)
+      .emplace(0, ActionTesting::MockDistributedObject<obs_writer>{});
+
+  // Specific IDs have no significance, just need different IDs.
+  const std::vector<ElementId<2>> element_ids{{1, {{{1, 0}, {1, 0}}}},
+                                              {1, {{{1, 0}, {2, 2}}}},
+                                              {0, {{{1, 0}, {1, 0}}}}};
+  for (const auto& id : element_ids) {
+    tuples::get<ElementMockDistributedObjectsTag>(dist_objects)
+        .emplace(ElementIndex<2>{id},
+
+                 ActionTesting::MockDistributedObject<element_comp>{
+                     db::create<element_comp::simple_tags,
+                                element_comp::compute_tags>(
+                         LinearSolver::IterationId{1},
+                         Mesh<2>{{{3, 2}},
+                                 Spectral::Basis::Legendre,
+                                 Spectral::Quadrature::GaussLobatto},
+                         Scalar<DataVector>{{{{1., 2., 3., 4., 5., 6.}}}},
+                         tnsr::I<DataVector, 2, Frame::Inertial>{
+                             {{{0., 1., 2., 3., 4., 5.},
+                               {-1., -2., -3., -4., -5., -6.}}}})});
+  }
+
+  tuples::TaggedTuple<AnalyticSolutionTag,
+                      observers::OptionTags::ReductionFileName,
+                      observers::OptionTags::VolumeFileName>
+      cache_data{};
+  const auto& reduction_file_name =
+      tuples::get<observers::OptionTags::ReductionFileName>(cache_data) =
+          "./Unit.Elliptic.Systems.Poisson.Actions_ReductionData";
+  const auto& volume_file_name =
+      tuples::get<observers::OptionTags::VolumeFileName>(cache_data) =
+          "./Unit.Elliptic.Systems.Poisson.Actions_VolumeData";
+  ActionTesting::MockRuntimeSystem<Metavariables> runner{
+      cache_data, std::move(dist_objects)};
+
+  runner.simple_action<obs_component,
+                       observers::Actions::Initialize<Metavariables>>(0);
+  runner.simple_action<obs_writer,
+                       observers::Actions::InitializeWriter<Metavariables>>(0);
+
+  // Register elements
+  for (const auto& id : element_ids) {
+    runner.simple_action<element_comp,
+                         observers::Actions::RegisterWithObservers<
+                             observers::TypeOfObservation::ReductionAndVolume>>(
+        id, 0);
+    // Invoke the simple_action RegisterSenderWithSelf that was called on the
+    // observer component by the RegisterWithObservers action.
+    runner.invoke_queued_simple_action<obs_component>(0);
+  }
+
+  const std::string reduction_h5_file_name = reduction_file_name + ".h5";
```

Comment:
is it possible that these file checks can be moved to just before the `obs_writer` invocations and merge the for loops?

Reply:
No I don't think the for loops can be merged. I assume the registration needs to be done before the elements start contributing. @nilsdeppe is that correct?
In any case, I don't think that change would actually improve this test, since it does not reflect the order in which the actions would be invoked in the executables and isn't important for performance.

Reply:
That's fine, then. I'd prefer the test to be as similar in structure to its eventual use case as possible, so given that this is the intended pattern for calling code, it's better left as is.

Reply:
Yes, registration must complete before anything is sent.